pkg/queue/manager.go

```diff
@@ -547,3 +594,90 @@ func (m *Manager) reportPendingWorkloads(cqName string, cq ClusterQueue) {
 	}
 	metrics.ReportPendingWorkloads(cqName, active, inadmissible)
 }
+
+func (m *Manager) Start(ctx context.Context) error {
+	defer m.queue.ShutDown()
+
+	queueVisibilityUpdateInterval := time.Duration(m.queueVisibilityUpdateInterval) * time.Second
+
+	for i := 0; i < snapshotWorkers; i++ {
+		go wait.UntilWithContext(ctx, m.takeSnapshot, queueVisibilityUpdateInterval)
+	}
+
+	go wait.UntilWithContext(ctx, m.enqueueSnapshotDelayed, queueVisibilityUpdateInterval)
+
+	<-ctx.Done()
+
+	return nil
+}
+
+func (m *Manager) enqueueSnapshotDelayed(ctx context.Context) {
+	m.RLock()
+	defer m.RUnlock()
+	for cq := range m.clusterQueues {
+		m.queue.Add(cq)
+	}
+}
+
+func (m *Manager) takeSnapshot(ctx context.Context) {
+	for m.processNextSnapshot(ctx) {
+	}
+}
+
+func (m *Manager) processNextSnapshot(ctx context.Context) bool {
+	m.wm.Lock()
+	defer m.wm.Unlock()
+
+	log := ctrl.LoggerFrom(ctx).WithName("processNextSnapshot")
+
+	key, quit := m.queue.Get()
+	if quit {
+		return false
+	}
+
+	startTime := time.Now()
+	defer func() {
+		log.V(4).Info("Finished snapshot job", "key", key, "elapsed", time.Since(startTime))
+	}()
+
+	cq := m.extractClusterQueue(key)
+	if cq == nil {
+		return false
+	}
+
+	defer m.queue.Done(key)
+
+	workloads := make([]kueue.ClusterQueuePendingWorkload, 0)
+	if elements, ok := cq.Snapshot(m.queueVisibilityClusterQueuesMaxCount); ok {
```

Comment:
IIUC taking the snapshot here is only protected by the m.wm.Lock, but not by the main m lock. I think this is fine, but I think we need another RW lock at the `clusterQueueBase` level to make sure the `heap` and `inadmissibleWorkloads` are in sync, without a need to take the lock for the entire manager.

@alculquicondor any thoughts?

Reply:
added rw lock in `clusterQueueBase`
https://github.com/kubernetes-sigs/kueue/pull/1069/files#diff-cc86534aa3b4b0d23e70ab9cefdd6fb23316d546052cac66fadecf7a53a69e40R229

Reply:
The write lock should cover also other functions that modify the heap or the inadmissibleWorkloads, so that the structure is consistent on reading. In particular, `QueueInadmissibleWorkloads` and `PushOrUpdate` need the write lock.